File: src/create_test_cloud.cpp
/**
 * @file create_test_cloud.cpp
 * @brief Creates a synthetic point cloud with intentional outliers for testing
 * 
 * This program generates a point cloud with two distinct distributions:
 * 1. A core set of points within a normal distribution range
 * 2. A smaller set of outlier points that are positioned farther from the center
 * 
 * The resulting point cloud is saved as a PCD file that can be used to test
 * outlier removal algorithms like StatisticalOutlierRemoval.
 * 
 * @author PCL Filtering Project
 * @date July 2025
 */

#include <iostream>
#include <pcl/io/pcd_io.h>
#include <pcl/point_types.h>
#include <random>

/**
 * @brief Main function for generating a test point cloud with outliers
 * 
 * Creates a point cloud with 1000 normal points and 50 outlier points.
 * Normal points are distributed within [-0.5, 0.5] range in all dimensions.
 * Outlier points are distributed within [-1.5, 1.5] range, making them
 * statistically distant from the core points.
 * 
 * @return int Exit code
 */
int main() {
    // Create point cloud object
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
    
    // Set basic parameters
    cloud->width = 1050;   // Total number of points (1000 normal + 50 outliers)
    cloud->height = 1;     // Unorganized point cloud (not a structured grid)
    cloud->is_dense = false; // May contain invalid points (NaN/Inf)
    cloud->points.resize(cloud->width * cloud->height);
    
    /**
     * Random Point Generation
     * 
     * We use two different distributions to create a point cloud with clear outliers:
     * 1. Normal points: Uniformly distributed in [-0.5, 0.5] range
     * 2. Outlier points: Uniformly distributed in [-1.5, 1.5] range
     * 
     * This creates a point cloud where the outliers are statistically distant from
     * the core points, making them good candidates for removal by the
     * StatisticalOutlierRemoval filter.
     */
    // Random number generators
    std::random_device rd;  // Used to obtain a seed for the random number engine
    std::mt19937 gen(rd()); // Standard mersenne_twister_engine seeded with rd()
    
    // Distribution for normal points (core cluster)
    std::uniform_real_distribution<> normal_dist(-0.5, 0.5);
    
    // Distribution for outlier points (wider range = farther from core)
    std::uniform_real_distribution<> outlier_dist(-1.5, 1.5);
    
    // Generate 1000 normal points that form the core of the point cloud
    for (size_t i = 0; i < 1000; ++i) {
        cloud->points[i].x = normal_dist(gen);
        cloud->points[i].y = normal_dist(gen);
        cloud->points[i].z = normal_dist(gen);
    }
    
    // Generate 50 outlier points that are statistically distant from the core
    for (size_t i = 1000; i < 1050; ++i) {
        cloud->points[i].x = outlier_dist(gen);
        cloud->points[i].y = outlier_dist(gen);
        cloud->points[i].z = outlier_dist(gen);
    }
    
    std::cout << "Generated point cloud with " << cloud->size() 
              << " points (1000 normal + 50 outliers)" << std::endl;
    
    // Save the point cloud
    pcl::io::savePCDFileASCII("input_cloud.pcd", *cloud);
    std::cout << "Saved point cloud to 'input_cloud.pcd'" << std::endl;
    
    return 0;
}

File: create_test_cloud.py
"""
Point Cloud Generator for Testing Statistical Outlier Removal

This script creates a synthetic point cloud with intentional outliers for testing
filtering algorithms. It uses Open3D to generate and visualize the point cloud.

The generated point cloud consists of:
1. 1000 normal points distributed within a cube centered at the origin
2. 50 outlier points distributed farther from the center

The resulting point cloud is saved as a PCD file that can be used with PCL.

Author: PCL Filtering Project
Date: July 2025
"""

import numpy as np
import open3d as o3d

def create_test_point_cloud():
    """
    Creates a test point cloud with normal points and outliers.
    
    The function generates two sets of points:
    - Normal points: 1000 points uniformly distributed in [-0.5, 0.5] range
    - Outlier points: 50 points distributed in a wider range (3x the normal range)
    
    Returns:
        o3d.geometry.PointCloud: The generated point cloud with outliers
    """
    # Create 1000 normal points in 3D space, centered around origin
    # These points form the core of the point cloud
    points = np.random.rand(1000, 3) - 0.5  # Range: [-0.5, 0.5]
    
    # Create 50 outlier points that are farther from the center
    # The factor of 3 makes these points statistically distant from the core
    outliers = (np.random.rand(50, 3) - 0.5) * 3  # Range: [-1.5, 1.5]
    
    # Combine normal points and outliers into a single array
    all_points = np.vstack([points, outliers])
    
    # Create Open3D point cloud object
    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(all_points)
    
    return pcd

# Generate the point cloud
point_cloud = create_test_point_cloud()

# Save to PCD file format (compatible with PCL)
o3d.io.write_point_cloud("input_cloud.pcd", point_cloud)
print("Created input_cloud.pcd with 1050 points (including 50 outliers)")

# Visualize the point cloud (optional)
# This will open a window showing the generated point cloud
# You can rotate, zoom, and pan using the mouse
o3d.visualization.draw_geometries([point_cloud], 
                                 window_name="Point Cloud with Outliers",
                                 width=800, 
                                 height=600)
